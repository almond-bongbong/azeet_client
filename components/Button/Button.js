import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import SimpleLoader from 'components/Loader/SimpleLoader';
import styled, { css } from 'styled-components';
import { red, blue } from 'style/mixin';

const BasicButtonStyle = styled.div`
  display: block;
  
  ${props => props.inline && css`
    display: inline-block;
    width: auto;
  `};

  button, a {
    display: block;
    width: ${props => `${props.width !== 'auto' ? `${props.width}px` : '100%'}`};
    min-width: 80px;
    height: ${props => props.height}px;
    padding: 0 20px;
    border: 1px solid #ddd;
    border-radius: 2px;
    outline: 0;
    font-size: 13px;
    cursor: pointer;
    text-align: center;
    text-decoration: none;
    line-height: ${props => props.height}px;
    transition: box-shadow .2s;
  
    &:hover {
      box-shadow: 1px 1px 3px rgba(0,0,0,0.2);
    }
    
    &:active {
      box-shadow: 0 0 2px rgba(0,0,0,0.1);
    }
    
    & span {
      position: relative;
    }
    
    ${props => props.theme === 'white' && css`
      background-color: #fff;
      border: 1px solid #ddd;
      color: #666;
    `};
    
    ${props => props.theme === 'gray' && css`
      background-color: #666;
      border: 1px solid #666;
      color: #fff;
    `}
    
    ${props => props.theme === 'red' && css`
      background-color: ${red};
      border: 1px solid ${red};
      color: #fff;
    `};
    
    ${props => props.theme === 'blue' && css`
      background-color: ${blue};
      border: 1px solid ${blue};
      color: #fff;
    `}
  }
`;

const Button = ({
  text, theme, width, height, inline, loading, onClick,
}) => {
  const loaderTheme = useMemo(() => (theme === 'white' ? 'red' : 'white'), [theme]);
  const loaderSize = useMemo(() => height / 2, [height]);

  return (
    <BasicButtonStyle data-testid="button-wrap" theme={theme} width={width} height={height} inline={inline}>
      {loading ? (
        <button type="button" title="로딩중"><SimpleLoader theme={loaderTheme} size={loaderSize} /></button>
      ) : (
        <button type="button" onClick={onClick}><span>{text}</span></button>
      )}
    </BasicButtonStyle>
  );
};

Button.propTypes = {
  text: PropTypes.string,
  inline: PropTypes.bool,
  theme: PropTypes.oneOf(['white', 'gray', 'red', 'blue']),
  width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  height: PropTypes.oneOf([30, 40, 50]),
  loading: PropTypes.bool,
  onClick: PropTypes.func,
};

Button.defaultProps = {
  text: '버튼',
  theme: 'white',
  width: 'auto',
  height: 40,
  inline: true,
  loading: false,
  onClick: undefined,
};

export default Button;
