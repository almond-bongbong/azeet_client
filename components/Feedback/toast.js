import React from 'react';
import ReactDOM from 'react-dom';
import { addRootElement, createElement } from 'lib/generateElement';
import styled from 'styled-components';
import ReactCSSTransitionGroup from 'react-addons-css-transition-group';

const toastContainer = document.getElementById('toast_container');
if (!toastContainer) addRootElement(createElement('toast_container'));
const toastComponentList = [];

const ToastStyle = styled.div`
  position: fixed;
  bottom: 20px;
  right: 30px;
  z-index: 1000;
  transform: translateZ(0);
  
  & .toast-enter {
    opacity: 0;
    height: 0;
    transform: translateY(10px);
  }
  & .toast-enter-active {
    opacity: 1;
    height: 41px;
    transform: translateY(0px);
  }
  & .toast-leave {
    opacity: 1;
    transform: translateY(0px);
  }
  & .toast-leave-active {
    opacity: 0;
    transform: translateY(-10px);
  }
  
  & .wrap {
    margin-bottom: 10px;
    text-align: right;
    font-size: 0;
    transition: height .3s, opacity .3s, transform .3s;
  }
  
  & .message {
    display: inline-block;
    min-width: 280px;
    overflow: hidden;
    height: 41px;
    padding: 10px 20px;
    background: rgba(0,0,10,0.4);
    box-shadow: 1px 2px 3px rgba(0,0,0,0.2);
    border-radius: 2px;
    color: #fff;
    font-size: 13px;
    white-space: nowrap;
    text-align: center;
  }
`;

const renderDOM = () => {
  const container = document.getElementById('toast_container');
  ReactDOM.render(
    <ToastStyle>
      <ReactCSSTransitionGroup
        transitionName="toast"
        transitionEnterTimeout={300}
        transitionLeaveTimeout={300}
      >
        {toastComponentList.map(t => t.component)}
      </ReactCSSTransitionGroup>
    </ToastStyle>,
    container,
  );
};

const toast = (message, time = 3000) => {
  renderDOM();

  const id = Date.now();
  toastComponentList.push({
    id,
    component: (
      <div key={id} className="wrap">
        <div className="message">{message}</div>
      </div>
    ),
  });

  renderDOM();
  setTimeout(() => {
    const index = toastComponentList.findIndex(t => t.id === id);
    toastComponentList.splice(index, 1);
    renderDOM();
  }, time);
};

export default toast;
