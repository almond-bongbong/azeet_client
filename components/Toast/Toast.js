import React from 'react';
import { useSelector } from 'react-redux';
import ReactCSSTransitionGroup from 'react-addons-css-transition-group';
import styled from 'styled-components';

const ToastStyle = styled.div`
  position: fixed;
  bottom: 20px;
  right: 30px;
  z-index: 1000;
  transform: translateZ(0);
  
  & .toast-enter {
    opacity: 0;
    height: 0;
    transform: translateY(10px);
  }
  & .toast-enter-active {
    opacity: 1;
    height: 41px;
    transform: translateY(0px);
  }
  & .toast-leave {
    opacity: 1;
    transform: translateY(0px);
  }
  & .toast-leave-active {
    opacity: 0;
    transform: translateY(-10px);
  }
  
  & .wrap {
    margin-bottom: 10px;
    text-align: right;
    font-size: 0;
    transition: height .3s, opacity .3s, transform .3s;
  }
  
  & .message {
    display: inline-block;
    min-width: 280px;
    overflow: hidden;
    height: 41px;
    padding: 10px 20px;
    background: rgba(0,0,10,0.4);
    box-shadow: 1px 2px 3px rgba(0,0,0,0.2);
    border-radius: 2px;
    color: #fff;
    font-size: 13px;
    white-space: nowrap;
    text-align: center;
  }
`;

const Toast = () => {
  const toast = useSelector(state => state.toast);

  return (
    <ToastStyle>
      <ReactCSSTransitionGroup
        transitionName="toast"
        transitionEnterTimeout={300}
        transitionLeaveTimeout={300}
      >
        {toast.map(t => (
          <div key={t.id} className="wrap">
            <div className="message">{t.message}</div>
          </div>
        ))}
      </ReactCSSTransitionGroup>
    </ToastStyle>
  );
};

export default Toast;
