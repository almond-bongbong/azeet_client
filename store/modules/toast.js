import { handleActions, createAction } from 'redux-actions';
import produce from 'immer';

export const types = {
  TOAST: 'toast/TOAST',
  DEQUEUE: 'toast/DEQUEUE',
};

export const toastActions = {
  toast: createAction(types.TOAST),
  dequeue: createAction(types.DEQUEUE),
};

const initialState = [];

export default handleActions({
  [types.TOAST]: (state, { payload }) => produce(state, (draft) => {
    draft.push({ id: draft.length, message: payload });
  }),
  [types.DEQUEUE]: state => produce(state, (draft) => {
    draft.shift();
  }),
}, initialState);
