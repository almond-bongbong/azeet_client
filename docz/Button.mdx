---
name: Button
---

import { Playground, Props } from 'docz'
import Button from 'components/Button/Button'

# Button

<Props of={Button} />

## basic usage

<Playground>
  <Button>Button</Button>
</Playground>

## inline

### true

<Playground>
  <Button theme="red">버튼</Button>
</Playground>

### false

<Playground>
  <Button theme="red" inline={false}>버튼</Button>
</Playground>

## theme

### white

<Playground>
  <Button theme="white">취소</Button>
</Playground>

### gray

<Playground>
  <Button theme="gray">취소</Button>
</Playground>

### red

<Playground>
  <Button theme="red">버튼</Button>
</Playground>

### blue

<Playground>
  <Button theme="blue">버튼</Button>
</Playground>

## width

<Playground>
  <Button width={120} theme="white">작은거</Button>
</Playground>

## height

### 30

<Playground>
  <Button height={30} theme="white">작은거</Button>
</Playground>

### 40

<Playground>
  <Button height={40} theme="white">보통</Button>
</Playground>

### 50

<Playground>
  <Button height={50} theme="white">큰거</Button>
</Playground>

## loading

<Playground>
  <Button loading={true}>로딩</Button>
</Playground>

## event

### onClick(event)

<Playground>
  <Button theme="white" onClick={e => alert('클릭')}>클릭!</Button>
</Playground>