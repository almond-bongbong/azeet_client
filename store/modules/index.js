import { combineReducers } from 'redux';
import alert from './alert';
import confirm from './confirm';
import toast from './toast';
import auth from './auth';
import bookmark from './bookmark';
import pending from './pending';

export default combineReducers({
  alert,
  confirm,
  toast,
  auth,
  bookmark,
  pending,
});
