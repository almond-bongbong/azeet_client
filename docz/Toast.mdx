---
name: Toast
---

import { Playground } from 'docz'
import { useDispatch } from 'react-redux'
import { toastActions } from 'store/modules/toast'
import Button from 'components/Button/Button'

# Toast

- 최상위 컴포넌트 내에 `<Toast />` 컴포넌트가 존재해야 합니다.
```
<div className="App">
  <Toast />
</div>

```

## Basic usage

- 아래와 같은 형식으로 toast를 호출 할 수 있습니다.
```
dispatch(toastActions.toast('hello toast!!'))
```

<Playground>
  {() => {
    const Test = () => {
      const dispatch = useDispatch();
      const openToast = () => dispatch(toastActions.toast('hello toast!!'));

      return <Button onClick={openToast}>alert!</Button>
    }

    return <Test />
  }}
</Playground>

## Time

- 노출 시간은 기본 3s 입니다.
- 다음과 같이 시간(ms)을 설정할 수 있습니다.

```
dispatch(toastActions.toast({ message: 'hello toast!!', time: 1000 }));
```

<Playground>
  {() => {
    const Test = () => {
      const dispatch = useDispatch();
      const openToast = () => dispatch(toastActions.toast({ message: 'hello toast!!', time: 1000 }));

      return <Button onClick={openToast}>alert!</Button>
    }

    return <Test />
  }}
</Playground>