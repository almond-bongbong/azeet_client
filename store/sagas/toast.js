import { all, fork, put, takeEvery, delay } from 'redux-saga/effects';
import { toastActions, types } from 'store/modules/toast';

function* toast() {
  yield delay(3000);
  yield put(toastActions.dequeue());
}

function* watchToast() {
  yield takeEvery(types.TOAST, toast);
}

export default function* toastSaga() {
  yield all([
    fork(watchToast),
  ]);
}
