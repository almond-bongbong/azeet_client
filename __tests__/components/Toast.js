import React from 'react';
import { fireEvent, wait } from '@testing-library/react';
import renderWithRedux from 'utils/renderWithRedux';
import Toast from 'components/Toast';
import Button from 'components/Button/Button';
import { toastActions } from 'store/modules/toast';

describe('Toast', () => {
  it('matches snapshot', () => {
    const utils = renderWithRedux(<Toast />);
    expect(utils.container).toMatchSnapshot();
  });

  it('calls toast popup', async () => {
    const { getByText, getAllByText, queryByText, store } = renderWithRedux(
      <>
        <Button text="click" onClick={() => store.dispatch(toastActions.toast('toast!'))} />
        <Toast />
      </>,
    );

    const button = getByText('click');
    fireEvent.click(button);
    fireEvent.click(button);
    expect(getAllByText('toast!').length).toBe(2);

    await wait(() => expect(queryByText('toast!')).toBeNull());
  });

  it('calls toast popup with time', async () => {
    const { getByText, queryByText, store } = renderWithRedux(
      <>
        <Button text="click" onClick={() => store.dispatch(toastActions.toast({ message: 'toast!', time: 500 }))} />
        <Toast />
      </>,
    );

    const button = getByText('click');
    fireEvent.click(button);
    getByText('toast!');

    await wait(() => expect(queryByText('toast!')).toBeNull());
  });
});
