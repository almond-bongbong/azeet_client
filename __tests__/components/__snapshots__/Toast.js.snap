// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast matches snapshot 1`] = `
<div>
  <div
    class="Toast__ToastStyle-sc-18p3u8j-0 jhiKCx"
  >
    <span />
  </div>
</div>
`;
